#' @importFrom GenomeInfoDb seqlevels
#' @importFrom GenomicAlignments seqlevelsInUse

#' @include gene_boundaries.R

#' @title Read BAM files as \code{\linkS4class[GenomicRanges]{GRanges}} object.
#'
#' Also sanitizes invalid chromosomes using \code{\link{limitChromosomes}}.
#'
#' @param bams A string vector of BAM file paths.
#' @return \code{\linkS4class[GenomicRanges]{GRanges}} S4 object combining all
#'     \code{bams}.
#' @seealso \code{\link{limitChromosomes}} for sanitization, and
#'     \code{\linkS4class[GenomicRanges]{GRanges}} for returned value.
#' @examples
#' gr <- readBams("foo.bam")
#' gr <- readBams(c("foo.bam", "baz/foo.bam"))
readBams <- function(bams, filter=TRUE) {
    ## Convert the input aligned reads into a bioconductor GRange object with
    ## non-standard chromosomes removed.
    grs <- lapply(bams, function(bam) {
        as(GenomicAlignments::readGAlignments(bam), "GRanges")
    })
    ## Combine all datasets.
    alignedReads <- unlist(GenomicRanges::GRangesList(grs))
    seqlevels(alignedReads) <- seqlevelsInUse(alignedReads)
    if (filter) {
        ## Remove non-standard "chrUn_*" and "*_random" chromosomes,
        ## otherwise groHMM::detectTranscripts() will fail with:
        ##
        ## Error in Fp[[i]] + 1 (from grohmm <- calibrate.R#140) :
        ## non-numeric argument to binary operator.
        filterChromosomes(alignedReads)
    }
}

#' @title Save and read HMM calibration to file generated by
#'     \code{\link[groHMM]{evaluateHMMInAnnotations}} \code{\link{data.frame}}.
#'
#' @name ioCal
#' @rdname ioCal
#' @param data A data.frame obtained from
#'     \code{\link[groHMM]{evaluateHMMInAnnotations}}.
#' @param file A string path.
#' @seealso \code{\link[groHMM]{evaluateHMMInAnnotations}} for
#' @examples
#' \dontrun{comparison <- groHMM::evaluateHMMInAnnotations(hmm$transcripts, consensus)}
#' \dontrun{hmm <- calibrate(reads=reads, LtProbB=LtProbB, UTS=UTS)}
#' fileCal <- "cell_line.grohmm.cal"
#' writeCal(ans, fileCal)
#' ## Now that we have a calibration for our cell line, we can use it in later R
#' ## sessions:
#' cal <- readCal(fileCal)
#' 
writeCal <- function(data, file) {
    ## Save HMM tuning paramters.
    write.table(data[, c("LtProbB", "UTS")],
                file=file,
                quote=FALSE,
                sep="\t",
                row.names=FALSE)
}

#' @rdname ioCal
readCal <- function(file) {
    ## Load HMM tuning parameters.
    read.table(file,
               header=TRUE)
}
